core/services/relay/evm/functions/contract_transmitter.go

```diff
@@ -0,0 +1,224 @@
+	return nil, nil
+}
+
+type contractTransmitter struct {
+	contractAddress     gethcommon.Address
+	contractABI         abi.ABI
+	transmitter         Transmitter
+	transmittedEventSig common.Hash
+	contractReader      contractReader
+	lp                  logpoller.LogPoller
+	lggr                logger.Logger
+	reportToEvmTxMeta   ReportToEthMetadata
+	contractVersion     uint32
+	reportCodec         encoding.ReportCodec
+}
+
+func transmitterFilterName(addr common.Address) string {
+	return logpoller.FilterName("OCR ContractTransmitter", addr.String())
+}
+
+func NewFunctionsContractTransmitter(
+	address gethcommon.Address,
+	caller contractReader,
+	contractABI abi.ABI,
+	transmitter Transmitter,
+	lp logpoller.LogPoller,
+	lggr logger.Logger,
+	reportToEvmTxMeta ReportToEthMetadata,
+	contractVersion uint32,
+) (*contractTransmitter, error) {
+	transmitted, ok := contractABI.Events["Transmitted"]
+	if !ok {
+		return nil, errors.New("invalid ABI, missing transmitted")
+	}
+
+	err := lp.RegisterFilter(logpoller.Filter{Name: transmitterFilterName(address), EventSigs: []common.Hash{transmitted.ID}, Addresses: []common.Address{address}})
+	if err != nil {
+		return nil, err
+	}
+	if reportToEvmTxMeta == nil {
+		reportToEvmTxMeta = reportToEvmTxMetaNoop
+	}
+	codec, err := encoding.NewReportCodec(contractVersion)
+	if err != nil {
+		return nil, err
+	}
+	return &contractTransmitter{
+		contractAddress:     address,
+		contractABI:         contractABI,
+		transmitter:         transmitter,
+		transmittedEventSig: transmitted.ID,
+		lp:                  lp,
+		contractReader:      caller,
+		lggr:                lggr.Named("OCRContractTransmitter"),
+		reportToEvmTxMeta:   reportToEvmTxMeta,
+		contractVersion:     contractVersion,
+		reportCodec:         codec,
+	}, nil
+}
+
+// Transmit sends the report to the on-chain smart contract's Transmit method.
+func (oc *contractTransmitter) Transmit(ctx context.Context, reportCtx ocrtypes.ReportContext, report ocrtypes.Report, signatures []ocrtypes.AttributedOnchainSignature) error {
+	var rs [][32]byte
+	var ss [][32]byte
+	var vs [32]byte
+	for i, as := range signatures {
+		r, s, v, err := evmutil.SplitSignature(as.Signature)
+		if err != nil {
+			panic("eventTransmit(ev): error in SplitSignature")
+		}
+		rs = append(rs, r)
+		ss = append(ss, s)
+		vs[i] = v
+	}
+	rawReportCtx := evmutil.RawReportContext(reportCtx)
+
+	txMeta, err := oc.reportToEvmTxMeta(report)
+	if err != nil {
+		oc.lggr.Warnw("failed to generate tx metadata for report", "err", err)
+	}
+
+	destinationContract := oc.contractAddress
+	if oc.contractVersion == 1 {
+		requests, err := oc.reportCodec.DecodeReport(report)
+		if err != nil {
+			return errors.Wrap(err, "DecodeReport failed")
+		}
+		// NOTE: this is incorrect if batch contains requests destined for different contracts
+		// and there is currently no way areound it until we get rid of batching
+		destinationContract.SetBytes(requests[0].CoordinatorContract)
```

Comment:
This seems like a big problem to me.  This means any in-flight requests will fail as soon as the upgrade occurs.  Also, how will Justin's `validateProposal` method work for the upgradeable contracts?
Here is how the flow needs to work from the perspective of the existing contracts:
1. An upgrade is proposed.  The CL node must receive this event and create a new `ContractTransmiter` which exists concurrently with the previous `ContractTransmitter`
2. We call `validateProposal` to send a test request using the new upgraded coordinator contract.  This must be handled by the new `ContractTransmitter`
3. All normal user requests must continue to use the previous `ContractTransmitter` until the upgrade is approved.
4. Once the upgrade is approved, we can remove the old `ContractTransmitter` and exclusively use the new `ContractTransmitter`

Given the flow above, the request must keep track of the coordinator contract from which it was emitted.  We cannot just set it here.

Reply:
Point (1) above is incorrect - there is only one transmitter that can handle multiple contracts.

Reply:
My misunderstanding was clarified during a sync with Bolek. Resolving.